File: app-ios-nativa/Fontaneria/Pruebas/PruebasFontaneria/PruebasGuardarEnlace.swift
import Dominio
import Foundation
import Testing

@testable import Fontaneria

@Suite("Guardar un enlace, desde donde sea")
struct PruebasGuardarEnlace {
    private func almacen() throws -> AlmacenLocal {
        try AlmacenLocal(ruta: ":memory:")
    }

    private let metadatos = MetadatosExtraidos(
        titulo: "Alternativas a Pocket",
        descripcion: "Una lista",
        imagenUrl: nil,
        tipo: .enlace
    )

    @Test("un enlace que no estaba se guarda y queda encolado")
    func enlaceNuevo() throws {
        let almacen = try almacen()

        let resultado = try GuardarEnlace.guardar(
            url: "https://ejemplo.com",
            etiquetas: ["ocio"],
            metadatos: metadatos,
            en: almacen,
            ahora: { 1_000 }
        )

        guard case .nuevo(let elemento) = resultado else {
            Issue.record("se esperaba un enlace nuevo")
            return
        }
        #expect(elemento.titulo == "Alternativas a Pocket")
        #expect(elemento.etiquetas == ["ocio"])
        #expect(try almacen.cargarPendientes()[elemento.id] != nil)
    }

    @Test("el enlace repetido actualiza el que había, como en el ordenador")
    func enlaceRepetido() throws {
        let almacen = try almacen()
        let primero = try GuardarEnlace.guardar(
            url: "https://ejemplo.com",
            etiquetas: ["ocio"],
            metadatos: metadatos,
            en: almacen,
            ahora: { 1_000 }
        )

        let segundo = try GuardarEnlace.guardar(
            url: "https://ejemplo.com",
            etiquetas: ["trabajo"],
            metadatos: nil,
            en: almacen,
            ahora: { 2_000 }
        )

        guard case .actualizado(let elemento) = segundo else {
            Issue.record("se esperaba una actualización")
            return
        }
        #expect(elemento.id == primero.elemento.id)
        #expect(elemento.etiquetas == ["ocio", "trabajo"])
        // Sin metadatos nuevos se conserva el título: quedarse sin él por
        // compartir el mismo enlace sin cobertura sería perder información.
        #expect(elemento.titulo == "Alternativas a Pocket")
        #expect(try almacen.cargarTodos().count == 1)
    }

    @Test("un enlace borrado y vuelto a guardar no deja dos filas")
    func enlaceQueSeHabiaBorrado() throws {
        let almacen = try almacen()
        let primero = try GuardarEnlace.guardar(
            url: "https://ejemplo.com",
            etiquetas: [],
            metadatos: metadatos,
            en: almacen,
            ahora: { 1_000 }
        )
        try almacen.marcarPendiente(primero.elemento.marcadoComoBorrado(ahora: { 1_500 }))

        let segundo = try GuardarEnlace.guardar(
            url: "https://ejemplo.com",
            etiquetas: [],
            metadatos: metadatos,
            en: almacen,
            ahora: { 2_000 }
        )

        // El borrado no cuenta como repetido: la fila sigue ahí de lápida,
        // pero para quien guarda el enlace ya no existe.
        guard case .nuevo = segundo else {
            Issue.record("un enlace borrado tiene que volver como nuevo")
            return
        }
    }

    @Test("lo que se dice al guardar distingue los tres casos")
    func loQueSeDice() throws {
        let almacen = try almacen()
        let nuevo = try GuardarEnlace.guardar(
            url: "https://ejemplo.com",
            etiquetas: [],
            metadatos: metadatos,
            en: almacen,
            ahora: { 1_000 }
        )

        #expect(nuevo.anuncio(seComprobo: true) == "Guardado, Alternativas a Pocket")
        #expect(
            nuevo.anuncio(seComprobo: false)
                == "Guardado sin título, no se pudo comprobar la página"
        )

        let repetido = try GuardarEnlace.guardar(
            url: "https://ejemplo.com",
            etiquetas: [],
            metadatos: metadatos,
            en: almacen,
            ahora: { 2_000 }
        )
        #expect(repetido.anuncio(seComprobo: true) == "Actualizado, Alternativas a Pocket")
    }
}

File: app-android/fontaneria/src/test/kotlin/com/jmortizsilva/guardarenlaces/fontaneria/PruebasGuardarEnlace.kt
package com.jmortizsilva.guardarenlaces.fontaneria

import com.jmortizsilva.guardarenlaces.dominio.MetadatosExtraidos
import kotlin.test.AfterTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertIs
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

class PruebasGuardarEnlace {
    private val almacen = almacenDePrueba()
    private val metadatos =
        MetadatosExtraidos(titulo = "Alternativas a Pocket", descripcion = "Una lista")

    @AfterTest fun cerrar() = almacen.cerrar()

    @Test
    fun `un enlace que no estaba se guarda y queda encolado`() {
        val resultado =
            GuardarEnlace.guardar("https://ejemplo.com", listOf("ocio"), metadatos, almacen) {
                1_000
            }

        assertIs<GuardarEnlace.Resultado.Nuevo>(resultado)
        assertEquals("Alternativas a Pocket", resultado.elemento.titulo)
        assertEquals(listOf("ocio"), resultado.elemento.etiquetas)
        assertNotNull(almacen.cargarPendientes()[resultado.elemento.id])
    }

    @Test
    fun `el enlace repetido actualiza el que habia, como en el ordenador`() {
        val primero =
            GuardarEnlace.guardar("https://ejemplo.com", listOf("ocio"), metadatos, almacen) {
                1_000
            }

        val segundo =
            GuardarEnlace.guardar("https://ejemplo.com", listOf("trabajo"), null, almacen) { 2_000 }

        assertIs<GuardarEnlace.Resultado.Actualizado>(segundo)
        assertEquals(primero.elemento.id, segundo.elemento.id)
        assertEquals(listOf("ocio", "trabajo"), segundo.elemento.etiquetas)
        // Sin metadatos nuevos se conserva el título: quedarse sin él por compartir el mismo
        // enlace sin cobertura sería perder información.
        assertEquals("Alternativas a Pocket", segundo.elemento.titulo)
        assertEquals(1, almacen.cargarTodos().size)
    }

    @Test
    fun `un enlace borrado y vuelto a guardar vuelve como nuevo`() {
        val primero =
            GuardarEnlace.guardar("https://ejemplo.com", emptyList(), metadatos, almacen) { 1_000 }
        almacen.marcarPendiente(primero.elemento.marcadoComoBorrado { 1_500 })

        val segundo =
            GuardarEnlace.guardar("https://ejemplo.com", emptyList(), metadatos, almacen) { 2_000 }

        // El borrado no cuenta como repetido: la fila sigue ahí de lápida, pero para quien guarda
        // el enlace ya no existe.
        assertIs<GuardarEnlace.Resultado.Nuevo>(segundo)
    }

    @Test
    fun `lo que se dice al guardar distingue los tres casos`() {
        val nuevo =
            GuardarEnlace.guardar("https://ejemplo.com", emptyList(), metadatos, almacen) { 1_000 }

        assertEquals("Guardado, Alternativas a Pocket", nuevo.anuncio(seComprobo = true))
        assertEquals(
            "Guardado sin título, no se pudo comprobar la página",
            nuevo.anuncio(seComprobo = false),
        )

        val repetido =
            GuardarEnlace.guardar("https://ejemplo.com", emptyList(), metadatos, almacen) { 2_000 }
        assertEquals("Actualizado, Alternativas a Pocket", repetido.anuncio(seComprobo = true))
    }
}

class PruebasCrearEtiqueta {
    private val almacen = almacenDePrueba()

    @AfterTest fun cerrar() = almacen.cerrar()

    @Test
    fun `la etiqueta nueva queda guardada y encolada`() {
        val etiqueta = CrearEtiqueta.crear("ocio", emptyList(), almacen) { 1_000 }

        assertEquals("ocio", etiqueta?.nombre)
        assertTrue(almacen.cargarEtiquetasDefinidas().values.any { it.nombre == "ocio" })
        assertEquals(1, almacen.cargarEtiquetasPendientes().size)
    }

    @Test
    fun `los espacios de alrededor no cuentan`() {
        assertEquals("trabajo", CrearEtiqueta.crear("  trabajo  ", emptyList(), almacen)?.nombre)
    }

    @Test
    fun `una que ya existe no se duplica, y no es un error`() {
        assertNull(CrearEtiqueta.crear("ocio", listOf("ocio"), almacen))
        assertTrue(almacen.cargarEtiquetasDefinidas().isEmpty())
    }

    @Test
    fun `un nombre en blanco no crea nada`() {
        assertNull(CrearEtiqueta.crear("   ", emptyList(), almacen))
        assertTrue(almacen.cargarEtiquetasDefinidas().isEmpty())
    }
}
